Refactor TextExtraction support in WKWebView to use a promise instead of a Swift block

whsieh opened the pull request:
#### cb875ef24116b4a39bab32f2aec4bdb16ebdf16f
<pre>
Refactor TextExtraction support in WKWebView to use a promise instead of a Swift block
<a href="https://bugs.webkit.org/show_bug.cgi?id=270031">https://bugs.webkit.org/show_bug.cgi?id=270031</a>
<a href="https://rdar.apple.com/123542653">rdar://123542653</a>

Reviewed by Richard Robinson and Abrar Rahman Protyasha.

Refactor this code to use a dedicated object (`WKTextExtractionRequest`) used to represent an
asynchronously-fulfilled promise, rather than taking a completion handler directly. As it turns out,
completion handler blocks in Swift-only API end up getting passed as `nil` into Objective-C methods
expecting blocks, when using `perform(with:…:)` to invoke ObjC-exposed selectors.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtractionForSwift:]):

Wrap the existing call to request text extraction.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionItem.h:
* Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
(WKTextExtractionRequest.completionHandler):
(WKTextExtractionRequest.fulfill(_:)):

Canonical link: <a href="https://commits.webkit.org/275286@main">https://commits.webkit.org/275286@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d673a7b2d74b630b412dce742b4af39f8c9591b5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41426 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20440 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43804 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43992 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37518 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23545 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17770 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34244 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42000 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17370 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35672 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14905 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15086 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45345 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37616 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36995 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40744 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16241 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13322 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39147 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17860 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9283 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17914 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17504 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
LGTM!

rr-codes commented on Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
```diff
@@ -120,3 +120,19 @@ import Foundation
+    @objc public let rectInWebView: CGRect
+    private var completionHandler: ((WKTextExtractionItem?) -> Void)?
+
+    @objc public init(rectInWebView: CGRect, completionHandler:@escaping (WKTextExtractionItem?) -> Void) {
```
```suggestion
    @objc public init(rectInWebView: CGRect, _ completionHandler: @escaping (WKTextExtractionItem?) -> Void) {
```

whsieh replied:
Applied these changes, and updated the call site to pass in `completionHandler` sans label.

Thanks for the review!

whsieh committed:
Refactor TextExtraction support in WKWebView to use a promise instead of a Swift block
https://bugs.webkit.org/show_bug.cgi?id=270031
rdar://123542653

Reviewed by Richard Robinson and Abrar Rahman Protyasha.

Refactor this code to use a dedicated object (`WKTextExtractionRequest`) used to represent an
asynchronously-fulfilled promise, rather than taking a completion handler directly. As it turns out,
completion handler blocks in Swift-only API end up getting passed as `nil` into Objective-C methods
expecting blocks, when using `perform(with:…:)` to invoke ObjC-exposed selectors.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtractionForSwift:]):

Wrap the existing call to request text extraction.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionItem.h:
* Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
(WKTextExtractionRequest.completionHandler):
(WKTextExtractionRequest.fulfill(_:)):

Canonical link: https://commits.webkit.org/275286@main
